### ISSUE TITLE
Is it really asynchronous?

### ISSUE THREAD

**Issue by StarlkYT** [chronoxor/NetCoreServer#245]
The library's README suggests that NetCoreServer is a "[...] **asynchronous** [...] solution", but is it really?
I can't find any usage of `async Task` in the README, and even methods like `SendAsync` in the `TCP` classes are not asynchronous.
I'm lost.

**Comment by chronoxor** [comment]
Yes, it's async streaming IO, implemented with Socket.ConnectAsync/Socket.SendAsync/Socket.ReceiveAsync & SocketAsyncEventArgs. Not with TPL, that is slower for raw bytes streaming.

If you want TPL with async/await pattern you might be interested in [high level message protocol](https://github.com/chronoxor/NetCoreServer/#example-simple-protocol) over NetCoreServer. Its client allows to send requests with timeout and Task<Response> result. Examples are here - [ProtoClient](https://github.com/chronoxor/NetCoreServer/#example-simple-protocol) & [ProtoServer](https://github.com/chronoxor/NetCoreServer/tree/master/examples/ProtoServer)

**Comment by StarlkYT** [comment]
I'm confused, I don't see any usage of async Task? Even methods with the async suffix don't return a Task and not async at all.
I'm probably missing something.
https://github.com/chronoxor/NetCoreServer/blob/master/examples/ProtoServer/Program.cs#L19

**Comment by StarlkYT** [comment]
A better question to ask would be, what if I want to write asynchronous code in method overrides like the OnReceieve ones? Am I left with doing `async void`?

**Comment by chronoxor** [comment]
It could be done with TaskCompletionSource:

```
        public Task<Response> Request(Request value)
        {
                var source = new TaskCompletionSource<Response>();
                Task<Response> task = source.Task;

                DateTime current = DateTime.UtcNow;

                // Send the request message
                SendAsync(request.Serialize());
                return task;
        }
```

```
        public bool OnReceive(...)
        {
                    // Read response from the buffer...
                    ...
                    source.SetResult(response);
        }
```

**Comment by StarlkYT** [comment]
Thanks! But, that's really not the best way, is it? Why not provide async Task overridables?

**Comment by chronoxor** [comment]
Also you can always access to TcpSession/TcpClient Socket property and call for [Task< int > ReceiveAsync(buffer)](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.receiveasync?view=net-7.0#system-net-sockets-socket-receiveasync(system-arraysegment((system-byte)))) and [Task< int > SendAsync (buffer)](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.sendasync?view=net-7.0#system-net-sockets-socket-sendasync(system-readonlymemory((system-byte))-system-threading-cancellationtoken)) to read/write buffer in async/await style

**Comment by chronoxor** [comment]
This library is event-based, which is faster than task-based approach. It is based on .net Sockets with some buffer management stuff and sessions. If your mission is not performance, you might use Socket Task/async methods and make it work in TPL style.

**Comment by BristleconeStudio** [comment]
> 

You could always write your own library the way you think it should work, couldn't you? I mean, you certainly seem to be in the mindset that chronoxor is incapable of doing it correctly, so why not show off your skills to us with a new superior library, instead of arguing about it?

Just a thought.

**Comment by StarlkYT** [comment]
It is indeed a though, and a very negative one! I was just asking behind the design decisions because I was probably missing something. And to potentially spark some ideas at the end.

**Comment by BristleconeStudio** [comment]
I see your point of view. It's okay for you to offer what you perceive to be constructive criticism, yet for me to suggest you contribute by writing your own library as opposed to accusing and complaining is negative, and triggered you to close the thread. Noted.

Enjoy your continued journey of learning and personal growth, Bader. You show great potential based on your repos.
